src/persist-txn/src/txn_read.rs

```diff
@@ -748,29 +896,29 @@ mod tests {
         let mut cache = TxnsCache::expect_open(3, &txns).await;
         assert_eq!(cache.progress_exclusive, 4);
 
-        // We can answer exclusive queries at <= progress. The data shard isn't
-        // registered yet, so they error.
-        assert!(cache.data_listen_next(&d0, 3).is_err());
-        assert!(cache.data_listen_next(&d0, 4).is_err());
+        // We can answer exclusive queries at <= progress. The data shard is not
+        // registered yet, so we read it as a normal shard.
+        assert_eq!(cache.data_listen_next(&d0, 3), ReadDataTo(4));
+        assert_eq!(cache.data_listen_next(&d0, 4), ReadDataTo(4));
 
-        // Register a data shard. Registration advances the physical upper, so
-        // we're able to read at the register ts.
+        // Register a data shard. The paired snapshot would advance the physical
+        // upper, so we're able to read at the register ts.
         cache.push_register(d0, 4, 1);
         cache.progress_exclusive = 5;
-        assert_eq!(cache.data_listen_next(&d0, 4), Ok(ReadDataTo(5)));
-        assert_eq!(cache.data_listen_next(&d0, 5), Ok(WaitForTxnsProgress));
+        // assert_eq!(cache.data_listen_next(&d0, 4), ReadDataTo(5));
```

Comment:
Just to get ahead of this, I think the `data_listen_next` test basically needs a rewrite to make sure it's hitting all the most interesting edge cases, but I'd like to hold off until the compaction PR to avoid having to do that twice